reconcile/utils/oc.py

```diff
@@ -544,12 +545,18 @@ def recycle_pods(self, dry_run, namespace, dep_kind, dep_resource):
                     stdin = json.dumps(obj, sort_keys=True)
                     self._run(cmd, stdin=stdin, apply=True)
 
-    def get_obj_root_owner(self, ns, obj):
+    def get_obj_root_owner(self, ns, obj, allow_not_found=False):
```

Comment:
If I understand correctly this function, it will:

1. return `obj` if no `ownerReferences`
2. return `obj` if `ownerReferences` are all set to `false`
3. return `obj` if `controller` exists and `allow_not_found` is set to `true` but referenced object does not exist
4. throw an exception  if `controller` exists and `allow_not_found` is set to `false` but referenced object does not exist
5. recurse if `controller` is true and referenced object exists

1 and 2 are fine if we're recursing, but it looks weird if the original object has no owner references. Most likely that edge case should be controlled in the caller

3 is weird. We find that the owner of our pod is no longer there so we return the pod `obj`? Shouldn't we return `{}` to tell the caller that the owner has been deleted?

As a final comment an as a testimony to the minutes that I have been reading this, this function:

* should have a nice piece of documentation.
* deserves a test since it is a critical one and we're inside a class where we can mock methods  😍

Reply:
about 3 - it's also fine. if an object had a parent and the parent is gone, it's like the object doesn't have a parent anymore, so it's like 1/2.

i'll add docs and tests

Reply:
Not having a parent is a different situation than having a deleted parent. In your case the caller may want to do the same, but they are different situations :)

Reply:
the caller did specify `allow_not_found=True`, so it means they want this behavior.

Reply:
no, they want to avoid non-existing parent make the method throw an error 😂 

Anyway, I already got some tests and documentation, so I'll give you this ;-)